ahnlich/ai/src/engine/store.rs

```diff
@@ -100,28 +106,57 @@ impl AIStoreHandler {
     }
 
     /// Converts storeinput into a tuple of storekey and storevalue.
-    /// Fails if the type of storeinput does not match the store type
+    /// Fails if the store input type does not match the store index_type
     #[tracing::instrument(skip(self))]
     pub(crate) fn store_input_to_store_key_val(
         &self,
         store_name: &StoreName,
         store_input: StoreInput,
         store_value: &StoreValue,
+        preprocess_action: &PreprocessAction,
     ) -> Result<(StoreKey, StoreValue), AIProxyError> {
         let metadata_key = &*AHNLICH_AI_RESERVED_META_KEY;
         if store_value.contains_key(metadata_key) {
             return Err(AIProxyError::ReservedError(metadata_key.to_string()));
         }
         let store = self.get(store_name)?;
 
-        let store_key = store.model.model_ndarray(&store_input);
-        let metadata_value: MetadataValue = store_input.into();
+        let store_input_type: AIStoreInputTypes = store_input.clone().into();
+        let store_index_model_info = store.index_model.model_info();
+
+        if store_input_type != store_index_model_info.input_type {
+            return Err(AIProxyError::StoreSetTypeMismatchError {
+                store_index_model_type: store_index_model_info.input_type,
+                storeinput_type: store_input_type,
+            });
+        }
+
+        let metadata_value: MetadataValue = store_input.clone().into();
         let mut final_store_value: StdHashMap<MetadataKey, MetadataValue> =
             store_value.clone().into_iter().collect();
         final_store_value.insert(metadata_key.clone(), metadata_value);
+
+        let store_key =
+            self.create_store_key(store_input, &store.index_model, preprocess_action)?;
         return Ok((store_key, final_store_value));
     }
 
+    /// Converts storeinput into a tuple of storekey and storevalue.
+    /// Fails if the store input type does not match the store index_type
+    #[tracing::instrument(skip(self))]
+    pub(crate) fn create_store_key(
+        &self,
+        store_input: StoreInput,
+        index_store_model: &AIModel,
+        preprocess_action: &PreprocessAction,
+    ) -> Result<StoreKey, AIProxyError> {
+        let mut store_input = store_input;
+
+        self.process_store_input(preprocess_action, &mut store_input, index_store_model)?;
```

Comment:
Put a comment on this ... I personally find these implicit mutations of variables just to use them immediately after a bit icky... however that's a nit and this isn't the final form especially because what the preprocess actions will return is actually some new enum containing processed string tokens or resized image content

Reply:
```rust
// allowed_input is a method that effectively tells us either the max amount of string tokens this model
// accepts as inputs or the image size it accepts as input
let preprocessed_input = self.preprocess_input(action, store_input, index_store_model.allowed_input())?;
let store_key = index_store_model.model_ndarray(preprocessed_input);
```